Library/OcAppleKernelLib/PrelinkedContext.c

```diff
@@ -899,6 +909,129 @@ PrelinkedInjectComplete (
+
+  // Figure out the value of NumSymbolsInPrelinked and LengthOfPrelinkedSymbols
+  Link                     = GetFirstNode (&Context->PrelinkedKexts);
+  Kext                     = NULL;
+  NumSymbolsInPrelinked    = 0;
+  LengthOfPrelinkedSymbols = sizeof (LILU_PRELINKED_SYMBOLS_HEADER);
+  while (!IsNull (&Context->PrelinkedKexts, Link)) {
+    Kext = GET_PRELINKED_KEXT_FROM_LINK (Link);
+    if (Kext->LinkedSymbolTable != NULL) {
+      NumSymbolsInKext       = Kext->NumberOfSymbols;
+      NumSymbolsInPrelinked += NumSymbolsInKext;
+      Symbols                = Kext->LinkedSymbolTable;
+
+      SymbolsEnd = &Symbols[NumSymbolsInKext];
+      while (Symbols < SymbolsEnd) {
+        LengthOfPrelinkedSymbols += sizeof (LILU_PRELINKED_SYMBOLS_ENTRY) + Symbols->Length;
+        Symbols++;
+      }
+    }
+
+    Link = GetNextNode (&Context->PrelinkedKexts, Link);
+  }
+
+  // Allocate buffer and setup header
+  LiluPrelinkedSymbolsAddr = (UINTN)AllocateRuntimePool (LengthOfPrelinkedSymbols);
+  Buffer                   = (VOID *)LiluPrelinkedSymbolsAddr;
+  if (!Buffer) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  LiluHeader                  = (LILU_PRELINKED_SYMBOLS_HEADER *)Buffer;
+  LiluHeader->Version         = 0;
+  LiluHeader->Size            = LengthOfPrelinkedSymbols;
+  LiluHeader->NumberOfSymbols = NumSymbolsInPrelinked;
+
+  // Fill in the entries
+  CurEntry = (LILU_PRELINKED_SYMBOLS_ENTRY *)(LiluHeader + 1);
+  Link     = GetFirstNode (&Context->PrelinkedKexts);
+  Kext     = NULL;
+  while (!IsNull (&Context->PrelinkedKexts, Link)) {
+    Kext = GET_PRELINKED_KEXT_FROM_LINK (Link);
+    if (Kext->LinkedSymbolTable != NULL) {
+      NumSymbolsInKext = Kext->NumberOfSymbols;
+      Symbols          = Kext->LinkedSymbolTable;
+
+      SymbolsEnd = &Symbols[NumSymbolsInKext];
+      while (Symbols < SymbolsEnd) {
+        CurEntry->EntryLength      = sizeof (LILU_PRELINKED_SYMBOLS_ENTRY) + Symbols->Length;
+        CurEntry->SymbolValue      = Symbols->Value - 0xFFFFFF8000100000ULL; // KERNEL_HIB_VADDR
+        CurEntry->SymbolNameLength = Symbols->Length;
+        CopyMem (&CurEntry->SymbolName, Symbols->Name, Symbols->Length + 1);
+
+        CurEntry = (LILU_PRELINKED_SYMBOLS_ENTRY *)(((UINT8 *)CurEntry) + CurEntry->EntryLength);
+        Symbols++;
+      }
+    }
+
+    Link = GetNextNode (&Context->PrelinkedKexts, Link);
+  }
+
+  LiluPrelinkedSymbolsAddr64 = (UINT64)LiluPrelinkedSymbolsAddr;
+  DEBUG ((
+    DEBUG_INFO,
+    "OCAK: Lilu prelinked symbols are stored at 0x%llx with a size of 0x%x bytes and %d symbols\n",
+    LiluPrelinkedSymbolsAddr64,
+    LengthOfPrelinkedSymbols,
+    NumSymbolsInPrelinked
+    ));
+
+  // Expose the physical address of the LILU_PRELINKED_SYMBOLS struct via an EFI variable
+  Status = OcSetSystemVariable (
+             OC_LILU_PRELINKED_SYMBOLS_ADDR_VARIABLE_NAME,
+             OPEN_CORE_NVRAM_ATTR,
+             8,
+             (void *)&LiluPrelinkedSymbolsAddr64,
```

Comment:
Redundant casts, also `void` is not allowed.

Reply:
Done